fmralign/pairwise_alignment.py

```diff
@@ -0,0 +1,310 @@
+        ward.fit(X)
+        labels = ward.labels_
+    return labels
+
+
+def generate_Xi_Yi(labels, X, Y):
+    unique_labels, counts = np.unique(labels, return_counts=True)
+    print(counts)
+    for k in range(len(unique_labels)):
+        label = unique_labels[k]
+        i = label == labels
+        if (k + 1) % 25 == 0:
+            print("Fitting parcel: " + str(k + 1) +
+                  "/" + str(len(unique_labels)))
+        yield X[i], Y[i]
+
+
+def fit_one_piece(X_i, Y_i, alignment_method):
+    """
+    Align source and target data in one piece i, X_i and Y_i, using alignment method and learn transformation to map X to Y.
+
+    Parameters
+    ----------
+    X_i: ndarray
+        Source data for piece i (shape : n_timeframes, n_i_voxels)
+    Y_i: ndarray
+        Target data for piece i (shape : n_timeframes, n_i_voxels)
+    alignment_method: string
+        algorithm used to perform alignment between X_i and Y_i can be taken
+        'identity', 'scaled_orthogonal', 'ridge_cv', 'permutation' or an instance of one of alignment classes provided in functional_alignment.alignment_methods
+
+    Returns
+    -------
+    alignment_algo
+        Instance of alignment estimator class fitted for X_i, Y_i
+    """
+
+    if alignment_method == 'identity':
+        alignment_algo = Identity()
+    elif alignment_method == 'scaled_orthogonal':
+        alignment_algo = ScaledOrthogonalAlignment()
+    elif alignment_method == 'ridge_cv':
+        alignment_algo = RidgeAlignment()
+    elif alignment_method == 'permutation':
+        alignment_algo = Hungarian()
+    elif alignment_method == 'optimal_transport':
+        alignment_algo = OptimalTransportAlignment()
+    elif isinstance(alignment_method, (Identity, ScaledOrthogonalAlignment, RidgeAlignment, Hungarian, OptimalTransportAlignment)):
+        alignment_algo = copy.deepcopy(alignment_method)
+    alignment_algo.fit(X_i.T, Y_i.T)
+
+    return alignment_algo
+
+
+def fit_one_parcellation(X_, Y_, alignment_method, mask, n_pieces, clustering_method, joint_clustering, clustering_index, mem,
+                         n_jobs=1, verbose=0):
+    """
+    Create one parcellation of n_pieces and align each source and target data in one piece i, X_i and Y_i, using alignment method and learn transformation to map X to Y.
+
+    Parameters
+    ----------
+    X_: ndarray
+        Source data (shape : n_timeframes, n_voxels)
+    Y_: ndarray
+        Target data (shape : n_timeframes, n_voxels)
+    alignment_method: string
+        algorithm used to perform alignment between each region of X_ and Y_
+    mask: Niimg-like object
+        Mask to be used on data.
+    clustering_method : string
+        method used to perform parcellation of data
+    joint_clustering : boolean
+        If True, clustering of voxels is performed jointly using X_ and Y_ signal. If false, it's performed only on X_
+    clustering_index : list of integers
```

Comment:
Currently, the default (and only possible) behavior is fixed in the fit() function :
```
rs = ShuffleSplit(n_splits=self.n_bags,
                          test_size=.8, random_state=0)

 Parallel(n_jobs=self.n_jobs, verbose=self.verbose)(
            delayed(fit_one_parcellation)(..., clustering_index, ...) 
            for clustering_index, _ in rs.split(Y_.T)
             )
```
This is a compromise between different needs (for fullbrain, parcellated, and bootstrapped cases)  we agreed on beforehand.
Do you think we need to add flexibility in this? It will add complexity to the API.
If you really find it unsatisfying for the point we're at, i'd rather tune the behavior to the specific usecase automatically (from n_pieces and n_bags) than add some more arguments.

Reply:
My impression is that we're over-engineering it. Why not simply start with a reasonable baseline and potentially revise it later is we feel it's too limited ?